crypto/params.c

```diff
@@ -0,0 +1,287 @@
+                   sz);
+    }
+    return 1;
+}
+
+static int OSSL_PARAM_set_int_common(const OSSL_PARAM *p, const char *key,
+                                     const void *val, size_t sz)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+    int neg, sign;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    sign = p->data_type == OSSL_PARAM_INTEGER;
+    if (!sign && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)
+        return 0;
+    if (p->buffer_size == sz) {         /* Fast path */
+        memcpy(p->buffer, val, sz);
+    } else if (p->buffer_size > sz) {   /* Widening */
+        const size_t eb = p->buffer_size - sz;
+
+        if (is_endian.little) {
+            neg = sign && (((unsigned char *)val)[p->buffer_size - 1]
+                           & 0x80);
+            memcpy(p->buffer, val, sz);
+            memset(((unsigned char *)p->buffer) + sz, neg ? 0xff : 0, eb);
+        } else {
+            neg = sign && (*(unsigned char *)val & 0x80);
+            memset(p->buffer, neg ? 0xff : 0, eb);
+            memcpy(((unsigned char *)p->buffer) + eb, val, sz);
+        }
+    } else {                            /* Narrowing */
+        if (is_endian.little)
+            memcpy(p->buffer, val, p->buffer_size);
+        else
+            memcpy(p->buffer, ((unsigned char *)val) + (sz - p->buffer_size),
+                   p->buffer_size);
+    }
+    if (p->return_size != NULL)
+        *p->return_size = sz;
+    return 1;
+}
+
+#define PARAM_INT(name, type) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return OSSL_PARAM_get_int_common(p, key, val, sizeof(*val)); \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, \
+                              const type *val) \
+    { \
+        return OSSL_PARAM_set_int_common(p, key, val, sizeof(*val)); \
+    }
+
+PARAM_INT(char, char)
+PARAM_INT(short, short int)
+PARAM_INT(int, int)
+PARAM_INT(long, long int)
+PARAM_INT(long_long, long long int)
+PARAM_INT(int8, int8_t)
+PARAM_INT(int16, int16_t)
+PARAM_INT(int32, int32_t)
+PARAM_INT(int64, int64_t)
+
+PARAM_INT(uchar, unsigned char)
+PARAM_INT(ushort, unsigned short int)
+PARAM_INT(uint, unsigned int)
+PARAM_INT(ulong, unsigned long int)
+PARAM_INT(ulong_long, unsigned long long int)
+PARAM_INT(uint8, uint8_t)
+PARAM_INT(uint16, uint16_t)
+PARAM_INT(uint32, uint32_t)
+PARAM_INT(uint64, uint64_t)
+PARAM_INT(size_t, size_t)
+
+int OSSL_PARAM_get_BN(const OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+    BIGNUM *b;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        if (is_endian.little)
+            b = BN_lebin2bn(p->buffer, p->buffer_size, *val);
```

Comment:
This assumes that little-endian uses the lebin encoding.  Are you sure that's true?  If so, why?

Reply:
That's actually what is specified if you read the manual on OSSL_PARAM.  "Native byte order"